tests/testthat/test-appMetadata.R

```diff
@@ -0,0 +1,291 @@
+
+test_that("leaves addPrimaryDoc unchanged or not a document", {
+  expect_equal(inferAppPrimaryDoc("foo.Rmd"), "foo.Rmd")
+  expect_equal(inferAppPrimaryDoc(NULL, appMode = "shiny"), NULL)
+  expect_equal(inferAppPrimaryDoc(NULL, appMode = "api"), NULL)
+})
+
+test_that("uses index file if present", {
+  files <- c("index.html", "index.Rmd", "a.html", "b.html", "a.Rmd", "b.Rmd")
+  expect_equal(inferAppPrimaryDoc(NULL, files, "static"), "index.html")
+  expect_equal(inferAppPrimaryDoc(NULL, files, "rmd-shiny"), "index.Rmd")
+})
+
+test_that("otherwise fails back to first file with matching extensions", {
+  files <- c("a.html", "b.html", "a.Rmd", "b.Rmd")
+  expect_equal(inferAppPrimaryDoc(NULL, files, "static"), "a.html")
+  expect_equal(inferAppPrimaryDoc(NULL, files, "rmd-shiny"), "a.Rmd")
+})
+
+test_that("errors if no files with needed extension", {
+  expect_snapshot(error = TRUE, {
+    inferAppPrimaryDoc(NULL, "a.R", "static")
+    inferAppPrimaryDoc(NULL, "a.R", "rmd-shiny")
+  })
+})
+
+# appHasParameters --------------------------------------------------------
+
+test_that("non-documents don't have parameters", {
+  dir <- local_temp_app(list("foo.R" = ""))
+
+  expect_false(appHasParameters(dir, "foo.R", "static"))
+  expect_false(appHasParameters(dir, "foo.R", "shiny"))
+})
+
+test_that("documents don't have parameters if part of a site", {
+  dir <- local_temp_app(list("index.Rmd" = c("---", "params: [1, 2]", "---")))
+
+  expect_false(appHasParameters(dir, "index.Rmd", "rmd-static", "site"))
+  expect_false(appHasParameters(dir, "index.Rmd", "qmd-shiny", "site"))
+})
+
+test_that("non-Rmd files don't have parameters", {
+  dir <- local_temp_app(list("app.r" = c("")))
+  expect_false(appHasParameters(dir, "app.R", "rmd-shiny"))
+})
+
+test_that("otherwise look at yaml metadata", {
+  dir <- local_temp_app(list("index.Rmd" = c("---", "params: [1, 2]", "---")))
+  expect_true(appHasParameters(dir, "index.Rmd", "rmd-shiny"))
+
+  dir <- local_temp_app(list("index.Rmd" = c("---", "params: ~", "---")))
+  expect_false(appHasParameters(dir, "index.Rmd", "rmd-shiny"))
+})
+
+
+# detectPythonInDocuments -------------------------------------------------
+
+test_that("dir without Rmds doesn't have have python", {
+  dir <- local_temp_app()
+  expect_false(detectPythonInDocuments(dir))
+
+  dir <- local_temp_app(list("foo.R" = ""))
+  expect_false(detectPythonInDocuments(dir))
+})
+
+test_that("Rmd or qmd with python chunk has python", {
+  dir <- local_temp_app(list("foo.qmd" = c("```{r}", "1+1", "````")))
+  expect_false(detectPythonInDocuments(dir))
+
+  dir <- local_temp_app(list("foo.Rmd" = c("```{python}", "1+1", "````")))
+  expect_true(detectPythonInDocuments(dir))
+
+  dir <- local_temp_app(list("foo.qmd" = c("```{python}", "1+1", "````")))
+  expect_true(detectPythonInDocuments(dir))
+})
+
+# quarto ------------------------------------------------------------------
+
+fakeQuartoMetadata <- function(version, engines) {
```

Comment:
This helper produces data for the existing `metadata` argument. I think the IDE can pass in other information through this optional argument. @toph-allen - do you have a reference to that?

Reply:
Here's where the Quarto `metadata` object comes from: https://github.com/quarto-dev/quarto-r/blob/08caf0f42504e74b7fc5382abed0820d5f415b3e/R/publish.R#L117-L121